File: openfeature-provider/go/confidence/remote_materialization_store.go
package confidence

import (
	"context"
	"fmt"
	"os"
	"strconv"
	"time"

	pb "github.com/spotify/confidence-resolver/openfeature-provider/go/confidence/internal/proto/resolverinternal"
	"google.golang.org/grpc/metadata"
)

const (
	defaultMaterializationReadTimeoutSeconds  = 2
	defaultMaterializationWriteTimeoutSeconds = 5
)

// remoteMaterializationStore is a MaterializationStore implementation that stores
// materialization data remotely via gRPC to the Confidence service.
//
// This implementation is useful when you want the Confidence service to manage
// materialization storage server-side rather than maintaining local state.
type remoteMaterializationStore struct {
	client       pb.InternalFlagLoggerServiceClient
	clientSecret string
}

// newRemoteMaterializationStore creates a new RemoteMaterializationStore with the given gRPC client.
func newRemoteMaterializationStore(client pb.InternalFlagLoggerServiceClient, clientSecret string) *remoteMaterializationStore {
	return &remoteMaterializationStore{
		client:       client,
		clientSecret: clientSecret,
	}
}

// getMaterializationReadTimeoutSeconds gets the materialization read timeout from environment or returns default
func getMaterializationReadTimeoutSeconds() time.Duration {
	if envVal := os.Getenv("CONFIDENCE_MATERIALIZATION_READ_TIMEOUT_SECONDS"); envVal != "" {
		if seconds, err := strconv.ParseInt(envVal, 10, 64); err == nil {
			return time.Duration(seconds) * time.Second
		}
	}
	return time.Duration(defaultMaterializationReadTimeoutSeconds) * time.Second
}

// getMaterializationWriteTimeoutSeconds gets the materialization write timeout from environment or returns default
func getMaterializationWriteTimeoutSeconds() time.Duration {
	if envVal := os.Getenv("CONFIDENCE_MATERIALIZATION_WRITE_TIMEOUT_SECONDS"); envVal != "" {
		if seconds, err := strconv.ParseInt(envVal, 10, 64); err == nil {
			return time.Duration(seconds) * time.Second
		}
	}
	return time.Duration(defaultMaterializationWriteTimeoutSeconds) * time.Second
}

// Read performs a batch read of materialization data from the remote service.
func (r *remoteMaterializationStore) Read(ctx context.Context, ops []ReadOp) ([]ReadResult, error) {
	if len(ops) == 0 {
		return []ReadResult{}, nil
	}

	// Add deadline to context if not already present
	if _, hasDeadline := ctx.Deadline(); !hasDeadline {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, getMaterializationReadTimeoutSeconds())
		defer cancel()
	}

	// Add authorization header with client secret
	md := metadata.Pairs("authorization", fmt.Sprintf("ClientSecret %s", r.clientSecret))
	ctx = metadata.NewOutgoingContext(ctx, md)

	// Convert ReadOps to proto format
	protoOps := make([]*pb.ReadOp, 0, len(ops))
	for _, op := range ops {
		protoOp, err := readOpToProto(op)
		if err != nil {
			return nil, fmt.Errorf("failed to convert read op: %w", err)
		}
		protoOps = append(protoOps, protoOp)
	}

	// Call gRPC service
	req := &pb.ReadOperationsRequest{
		Ops: protoOps,
	}

	resp, err := r.client.ReadMaterializedOperations(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("failed to read materialized operations: %w", err)
	}

	// Convert proto results to Go types
	results := make([]ReadResult, 0, len(resp.Results))
	for _, protoResult := range resp.Results {
		result, err := protoToReadResult(protoResult)
		if err != nil {
			return nil, fmt.Errorf("failed to convert read result: %w", err)
		}
		results = append(results, result)
	}

	return results, nil
}

// Write performs a batch write of materialization data to the remote service.
func (r *remoteMaterializationStore) Write(ctx context.Context, ops []WriteOp) error {
	if len(ops) == 0 {
		return nil
	}

	// Add deadline to context if not already present
	if _, hasDeadline := ctx.Deadline(); !hasDeadline {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, getMaterializationWriteTimeoutSeconds())
		defer cancel()
	}

	// Add authorization header with client secret
	md := metadata.Pairs("authorization", fmt.Sprintf("ClientSecret %s", r.clientSecret))
	ctx = metadata.NewOutgoingContext(ctx, md)

	// Convert WriteOps to proto format
	protoOps := make([]*pb.VariantData, 0, len(ops))
	for _, op := range ops {
		protoOp, err := writeOpToProto(op)
		if err != nil {
			return fmt.Errorf("failed to convert write op: %w", err)
		}
		protoOps = append(protoOps, protoOp)
	}

	// Call gRPC service
	req := &pb.WriteOperationsRequest{
		StoreVariantOp: protoOps,
	}

	_, err := r.client.WriteMaterializedOperations(ctx, req)
	if err != nil {
		return fmt.Errorf("failed to write materialized operations: %w", err)
	}

	return nil
}

// readOpToProto converts a Go ReadOp interface to a proto ReadOp
func readOpToProto(op ReadOp) (*pb.ReadOp, error) {
	switch v := op.(type) {
	case *ReadOpVariant:
		return &pb.ReadOp{
			Op: &pb.ReadOp_VariantReadOp{
				VariantReadOp: &pb.VariantReadOp{
					Unit:            v.Unit(),
					Materialization: v.Materialization(),
					Rule:            v.Rule(),
				},
			},
		}, nil
	case *ReadOpInclusion:
		return &pb.ReadOp{
			Op: &pb.ReadOp_InclusionReadOp{
				InclusionReadOp: &pb.InclusionReadOp{
					Unit:            v.Unit(),
					Materialization: v.Materialization(),
				},
			},
		}, nil
	default:
		return nil, fmt.Errorf("unknown read op type: %T", op)
	}
}

// protoToReadResult converts a proto ReadResult to a Go ReadResult interface
func protoToReadResult(protoResult *pb.ReadResult) (ReadResult, error) {
	if protoResult == nil {
		return nil, fmt.Errorf("read result is nil")
	}

	// Access the result field which contains the oneof
	if protoResult.Result == nil {
		return nil, fmt.Errorf("read result.result is nil")
	}

	switch v := protoResult.Result.(type) {
	case *pb.ReadResult_VariantResult:
		if v.VariantResult == nil {
			return nil, fmt.Errorf("variant result is nil")
		}
		variant := v.VariantResult.Variant
		return &ReadResultVariant{
			materialization: v.VariantResult.Materialization,
			unit:            v.VariantResult.Unit,
			rule:            v.VariantResult.Rule,
			variant:         &variant,
		}, nil
	case *pb.ReadResult_InclusionResult:
		if v.InclusionResult == nil {
			return nil, fmt.Errorf("inclusion result is nil")
		}
		return &ReadResultInclusion{
			materialization: v.InclusionResult.Materialization,
			unit:            v.InclusionResult.Unit,
			included:        v.InclusionResult.IsIncluded,
		}, nil
	default:
		return nil, fmt.Errorf("unknown read result type: %T", protoResult.Result)
	}
}

// writeOpToProto converts a Go WriteOp interface to a proto VariantData
func writeOpToProto(op WriteOp) (*pb.VariantData, error) {
	switch v := op.(type) {
	case *WriteOpVariant:
		return &pb.VariantData{
			Unit:            v.Unit(),
			Materialization: v.Materialization(),
			Rule:            v.Rule(),
			Variant:         v.Variant(),
		}, nil
	default:
		return nil, fmt.Errorf("unknown write op type: %T", op)
	}
}

File: openfeature-provider/java/src/main/java/com/spotify/confidence/RemoteMaterializationStore.java
package com.spotify.confidence;

import static com.spotify.confidence.GrpcUtil.createConfidenceChannel;

import com.google.common.annotations.VisibleForTesting;
import com.spotify.confidence.flags.resolver.v1.*;
import io.grpc.*;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * A {@link MaterializationStore} implementation that stores materialization data remotely via gRPC
 * to the Confidence service.
 *
 * <p>This implementation is useful when you want the Confidence service to manage materialization
 * storage server-side rather than maintaining local state. The Confidence service automatically
 * handles TTL management (90-day default) for stored materializations.
 *
 * <p><strong>Usage Example:</strong>
 *
 * <pre>{@code
 * String clientSecret = "your-application-client-secret";
 * MaterializationStore store = new RemoteMaterializationStore(clientSecret);
 *
 * // Use with OpenFeatureLocalResolveProvider
 * OpenFeatureLocalResolveProvider provider =
 *     new OpenFeatureLocalResolveProvider(config, clientSecret, store);
 * }</pre>
 *
 * <p><strong>Configuration:</strong> Timeouts can be configured via environment variables:
 *
 * <ul>
 *   <li>{@code CONFIDENCE_MATERIALIZATION_READ_TIMEOUT_SECONDS} - timeout for read operations
 *       (default: 2 seconds)
 *   <li>{@code CONFIDENCE_MATERIALIZATION_WRITE_TIMEOUT_SECONDS} - timeout for write operations
 *       (default: 5 seconds)
 * </ul>
 *
 * <p><strong>Thread Safety:</strong> This implementation is thread-safe and can handle concurrent
 * requests.
 *
 * @see MaterializationStore
 * @see UnsupportedMaterializationStore
 */
public class RemoteMaterializationStore implements MaterializationStore {
  private static final Logger logger = LoggerFactory.getLogger(RemoteMaterializationStore.class);
  private static final Duration DEFAULT_READ_TIMEOUT = Duration.ofSeconds(2);
  private static final Duration DEFAULT_WRITE_TIMEOUT = Duration.ofSeconds(5);

  private final InternalFlagLoggerServiceGrpc.InternalFlagLoggerServiceBlockingStub stub;
  private final Duration readTimeout;
  private final Duration writeTimeout;
  private final ExecutorService executor;
  private final boolean shouldShutdownExecutor;
  private ManagedChannel channel;

  /**
   * Creates a new RemoteMaterializationStore with the given client secret.
   *
   * <p>This constructor uses the default channel factory to create a gRPC connection to the
   * Confidence service.
   *
   * @param clientSecret the client secret for authentication with the Confidence service
   */
  public RemoteMaterializationStore(String clientSecret) {
    this(clientSecret, new DefaultChannelFactory());
  }

  /**
   * Creates a new RemoteMaterializationStore with a custom channel factory.
   *
   * <p>This constructor is useful for testing or when you need custom gRPC channel configuration.
   *
   * @param clientSecret the client secret for authentication with the Confidence service
   * @param channelFactory the factory to use for creating gRPC channels
   */
  public RemoteMaterializationStore(String clientSecret, ChannelFactory channelFactory) {
    this.stub = createAuthStub(channelFactory, clientSecret);
    this.readTimeout = getReadTimeout();
    this.writeTimeout = getWriteTimeout();
    this.executor = Executors.newCachedThreadPool();
    this.shouldShutdownExecutor = true;
  }

  @VisibleForTesting
  RemoteMaterializationStore(
      InternalFlagLoggerServiceGrpc.InternalFlagLoggerServiceBlockingStub stub,
      Duration readTimeout,
      Duration writeTimeout,
      ExecutorService executor) {
    this.stub = stub;
    this.readTimeout = readTimeout;
    this.writeTimeout = writeTimeout;
    this.executor = executor;
    this.shouldShutdownExecutor = false;
  }

  private InternalFlagLoggerServiceGrpc.InternalFlagLoggerServiceBlockingStub createAuthStub(
      ChannelFactory channelFactory, String clientSecret) {
    this.channel = createConfidenceChannel(channelFactory);
    return InternalFlagLoggerServiceGrpc.newBlockingStub(channel)
        .withInterceptors(
            new ClientInterceptor() {
              @Override
              public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
                  MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
                return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
                    next.newCall(method, callOptions)) {
                  @Override
                  public void start(Listener<RespT> responseListener, Metadata headers) {
                    headers.put(
                        Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER),
                        "ClientSecret " + clientSecret);
                    super.start(responseListener, headers);
                  }
                };
              }
            });
  }

  private static Duration getReadTimeout() {
    String envVal = System.getenv("CONFIDENCE_MATERIALIZATION_READ_TIMEOUT_SECONDS");
    if (envVal != null && !envVal.isEmpty()) {
      try {
        return Duration.ofSeconds(Long.parseLong(envVal));
      } catch (NumberFormatException e) {
        logger.warn(
            "Invalid CONFIDENCE_MATERIALIZATION_READ_TIMEOUT_SECONDS value: {}, using default",
            envVal);
      }
    }
    return DEFAULT_READ_TIMEOUT;
  }

  private static Duration getWriteTimeout() {
    String envVal = System.getenv("CONFIDENCE_MATERIALIZATION_WRITE_TIMEOUT_SECONDS");
    if (envVal != null && !envVal.isEmpty()) {
      try {
        return Duration.ofSeconds(Long.parseLong(envVal));
      } catch (NumberFormatException e) {
        logger.warn(
            "Invalid CONFIDENCE_MATERIALIZATION_WRITE_TIMEOUT_SECONDS value: {}, using default",
            envVal);
      }
    }
    return DEFAULT_WRITE_TIMEOUT;
  }

  @Override
  public CompletionStage<List<ReadResult>> read(List<? extends ReadOp> ops) {
    if (ops == null || ops.isEmpty()) {
      return CompletableFuture.completedFuture(Collections.emptyList());
    }

    return CompletableFuture.supplyAsync(
        () -> {
          try {
            // Convert ReadOps to proto format
            List<com.spotify.confidence.flags.resolver.v1.ReadOp> protoOps =
                new ArrayList<>(ops.size());
            for (ReadOp op : ops) {
              protoOps.add(readOpToProto(op));
            }

            // Build request
            ReadOperationsRequest request =
                ReadOperationsRequest.newBuilder().addAllOps(protoOps).build();

            // Make gRPC call with timeout
            ReadOperationsResult response =
                stub.withDeadlineAfter(readTimeout.toMillis(), TimeUnit.MILLISECONDS)
                    .readMaterializedOperations(request);

            // Convert proto results to Java types
            List<ReadResult> results = new ArrayList<>(response.getResultsCount());
            for (com.spotify.confidence.flags.resolver.v1.ReadResult protoResult :
                response.getResultsList()) {
              results.add(protoToReadResult(protoResult));
            }

            return results;
          } catch (StatusRuntimeException e) {
            logger.error("Failed to read materialized operations", e);
            throw new RuntimeException("Failed to read materialized operations", e);
          }
        },
        executor);
  }

  @Override
  public CompletionStage<Void> write(Set<? extends WriteOp> ops) {
    if (ops == null || ops.isEmpty()) {
      return CompletableFuture.completedFuture(null);
    }

    return CompletableFuture.runAsync(
        () -> {
          try {
            // Convert WriteOps to proto format
            List<VariantData> protoOps = new ArrayList<>(ops.size());
            for (WriteOp op : ops) {
              protoOps.add(writeOpToProto(op));
            }

            // Build request
            WriteOperationsRequest request =
                WriteOperationsRequest.newBuilder().addAllStoreVariantOp(protoOps).build();

            // Make gRPC call with timeout
            stub.withDeadlineAfter(writeTimeout.toMillis(), TimeUnit.MILLISECONDS)
                .writeMaterializedOperations(request);

          } catch (StatusRuntimeException e) {
            logger.error("Failed to write materialized operations", e);
            throw new RuntimeException("Failed to write materialized operations", e);
          }
        },
        executor);
  }

  public void shutdown() {
    this.channel.shutdown();
    if (shouldShutdownExecutor) {
      this.executor.shutdown();
    }
  }

  private com.spotify.confidence.flags.resolver.v1.ReadOp readOpToProto(ReadOp op) {
    com.spotify.confidence.flags.resolver.v1.ReadOp.Builder builder =
        com.spotify.confidence.flags.resolver.v1.ReadOp.newBuilder();

    if (op instanceof ReadOp.Variant variant) {
      builder.setVariantReadOp(
          VariantReadOp.newBuilder()
              .setUnit(variant.unit())
              .setMaterialization(variant.materialization())
              .setRule(variant.rule())
              .build());
    } else if (op instanceof ReadOp.Inclusion inclusion) {
      builder.setInclusionReadOp(
          InclusionReadOp.newBuilder()
              .setUnit(inclusion.unit())
              .setMaterialization(inclusion.materialization())
              .build());
    } else {
      throw new IllegalArgumentException("Unknown read op type: " + op.getClass().getName());
    }

    return builder.build();
  }

  private ReadResult protoToReadResult(com.spotify.confidence.flags.resolver.v1.ReadResult proto) {
    if (proto.hasVariantResult()) {
      VariantData variantData = proto.getVariantResult();
      Optional<String> variant =
          variantData.getVariant() != null && !variantData.getVariant().isEmpty()
              ? Optional.of(variantData.getVariant())
              : Optional.empty();
      return new ReadResult.Variant(
          variantData.getMaterialization(), variantData.getUnit(), variantData.getRule(), variant);
    } else if (proto.hasInclusionResult()) {
      InclusionData inclusionData = proto.getInclusionResult();
      return new ReadResult.Inclusion(
          inclusionData.getMaterialization(),
          inclusionData.getUnit(),
          inclusionData.getIsIncluded());
    } else {
      throw new IllegalArgumentException("Unknown read result type");
    }
  }

  private VariantData writeOpToProto(WriteOp op) {
    if (op instanceof WriteOp.Variant variant) {
      return VariantData.newBuilder()
          .setUnit(variant.unit())
          .setMaterialization(variant.materialization())
          .setRule(variant.rule())
          .setVariant(variant.variant())
          .build();
    } else {
      throw new IllegalArgumentException("Unknown write op type: " + op.getClass().getName());
    }
  }
}
